File: Factory/PizzaStore.CS/PizzaStore/Pizza.cs
namespace PizzaStore;


public abstract class Pizza
{
    public String Name { get; set; }
    public Dough Dough { get; set; }
    public Sauce Sauce { get; set; }
    public List<Veggies> Veggies { get; } = new();
    public Cheese Cheese { get; set; }
    public Pepperoni Pepperoni { get; set; }
    public Clams Clam { get; set; }

    public abstract void Prepare();

    public virtual void Bake()
    {
        Console.WriteLine("Bake for 25 minutes at 350");
    }

    public virtual void Cut()
    {
        Console.WriteLine("Cutting the pizza into diagonal slices");
    }

    public virtual void Box()
    {
        Console.WriteLine("Place pizza in official PizzaStore box");
    }
}

public class CheesePizza : Pizza
{
    private PizzaIngredientFactory IngredientFactory { get; }

    public CheesePizza(PizzaIngredientFactory ingredientFactory)
    {
        this.IngredientFactory = ingredientFactory;
    }

    public override void Prepare()
    {
        Console.WriteLine($"Preparing {this.Name}");
        this.Dough = this.IngredientFactory.CreateDough();
        this.Sauce = this.IngredientFactory.CreateSauce();
        this.Cheese = this.IngredientFactory.CreateCheese();
    }
}

public class ClamPizza : Pizza
{
    private PizzaIngredientFactory IngredientFactory { get; }

    public ClamPizza(PizzaIngredientFactory ingredientFactory)
    {
        this.IngredientFactory = ingredientFactory;
    }

    public override void Prepare()
    {
        Console.WriteLine($"Preparing {this.Name}");
        this.Dough = this.IngredientFactory.CreateDough();
        this.Sauce = this.IngredientFactory.CreateSauce();
        this.Cheese = this.IngredientFactory.CreateCheese();
        this.Clam = this.IngredientFactory.CreateClam();
    }
}

File: Factory/PizzaStore.Py/src/Pizza.py
from abc import ABC, abstractmethod

from . import Ingredients


class Pizza(ABC):
    name: str
    dough: Ingredients.Dough
    sauce: Ingredients.Sauce
    veggies: list[Ingredients.Veggies]
    cheese: Ingredients.Cheese
    pepperoni: Ingredients.Pepperoni
    clams: Ingredients.Clams

    @abstractmethod
    def prepare(self) -> None:...

    def bake(self) -> None:
        print("Bake for 25 minutes at 350")

    def cut(self) -> None:
        print("Cutting the pizza into diagonal slices")

    def box(self) -> None:
        print("Place pizza in official PizzaStore box")


class CheesePizza(Pizza):

    def __init__(self, ingredient_factory: Ingredients.PizzaIngredientFactory) -> None:
        self._ingredient_factory = ingredient_factory

    def prepare(self) -> None:
        print(f"Preparing {self.name}")
        self.dough = self._ingredient_factory.create_dough()
        self.sauce = self._ingredient_factory.create_sauce()
        self.cheese = self._ingredient_factory.create_cheese()


class ClamPizza(Pizza):

    def __init__(self, ingredient_factory: Ingredients.PizzaIngredientFactory) -> None:
        self._ingredient_factory = ingredient_factory

    def prepare(self) -> None:
        print(f"Preparing {self.Name}")
        self.dough = self._ngredient_factory.create_dough()
        self.sauce = self._ngredient_factory.create_sauce()
        self.cheese = self._ngredient_factory.create_cheese()
        self.clam = self._ngredient_factory.create_clam()